src/main/daml/Daml/Finance/Interface/Util/SyntheticKey.daml

```diff
@@ -0,0 +1,174 @@
+-- Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE AllowAmbiguousTypes #-}
+
+module Daml.Finance.Interface.Util.SyntheticKey where
+
+import DA.Assert ((===))
+import DA.Map qualified as M (Map, delete, empty, insert, lookup)
+import DA.Optional (fromSome)
+import DA.Text (sha256)
+import DA.Traversable qualified as T (mapA)
+import Daml.Control.Monad.Trans.Class
+import Daml.Control.Monad.Trans.State
+import Prelude hiding (lookup)
+
+-- | Type synonym for interface.
+type I = SyntheticKey
+
+-- | Type synonym for `View`.
+type V = View
+
+-- | An interface for synthetically keyed interfaces.
+interface SyntheticKey where
+  viewtype V
+
+-- | View for `I`.
+data View = View {}
+  deriving (Eq, Show)
+
+-- ============= Generic Index Table =============
+-- | Index table (used for looking up contract ids for keys).
+newtype IndexTable = IndexTable {getIndexTable : M.Map Text (ContractId I)} deriving (Eq, Ord, Show)
+
+-- HIDE
+instance Semigroup IndexTable where
+  (<>) a b = IndexTable $ getIndexTable a <> getIndexTable b
+
+-- HIDE
+instance Monoid IndexTable where
+  mempty = IndexTable M.empty
+  mconcat = foldl (<>) mempty
+
+-- | Insert.
+insert : (Show k, HasToInterface i I) => k -> ContractId i -> IndexTable -> IndexTable
+insert k cid idx =
+  IndexTable $ M.insert (sha256 $ show k) (toInterfaceContractId @I cid) (getIndexTable idx)
```

Comment:
Is there a reason why `show k` would not be sufficient as a key? hashing here makes the key non human readable

Reply:
The reason for hashing was to keep the key reasonably short (sha256 is about 64 characters), but we could as well use the `show k` value as the key for our purposes. I don't expect we would need to access the index table manually, except for debugging purposes.

Note that when a user needs to exercise a choice that requires an index table, they only need to include the contract IDs necessary for that action, not the entire table. If one choice calls another choice that only needs a subset of the table, it can pass on just that subset. Usually, we pass an `IndexTable` as an argument to look up corresponding contract IDs for keys we already know, so we don't reveal more information. However, in other cases, we might leak more information (especially if we provide entries which are not needed).